iwyu_test_util.py

```diff
@@ -73,12 +73,54 @@ class Features:
   # Parses the output of 'include-what-you-use -print-targets'.
   SUPPORTED_TARGET_RE = re.compile(r'^\s+([a-z0-9_-]+)\s* - .*$')
 
+  # Parses the output of 'include-what-you-use -dM'.
+  PREPROCESSOR_MACRO_RE = re.compile(r'^#define ([A-z_][A-z0-9_]*)[\s\(].*$')
+
   def __init__(self):
     # Maps all known features names to prerequisite predicates.
     self.known = {
-      'has-target': Features.HasTarget
+      'has-target': Features.HasTarget,
+      'using-cstdlib': Features.UsingCstdlib,
+      'using-cxxstdlib': Features.UsingCXXstdlib,
     }
 
+  def _GetPreprocessorMacros(lang, header):
+    iwyu = _GetIwyuPath()
+    txt = f'#include {header}'
+
+    result = subprocess.run([iwyu, '-dM', '-x', lang, '-'], capture_output=True,
```

Comment:
For one, I was thinking that we should probably use the same command-line as would be used for the test (the one built in `TestIwyuOnRelativeFile`), but I don't really see how to get that into here nicely.

But second, shouldn't the feature be more like "is <library> installed?", in which case the test can explicitly specify `-stdlib=libc++`, or whatever it needs. Now it seems the test suite runs or not depending on what happens to be the compiler default for the platform.

Or maybe both are necessary...?

Reply:
I did a little experiment with Clang in a clean docker environment:
```bash
$ apt install clang

$ head /usr/include/c++/11/cstddef 
// -*- C++ -*- forwarding header.

// Copyright (C) 1997-2021 Free Software Foundation, Inc.
//
// This file is part of the GNU ISO C++ Library.  This library is free
// software; you can redistribute it and/or modify it under the
// terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option)
// any later version.

# shows libstdc++ ^

$ echo "#include <cstddef>" | clang -stdlib=libstdc++ -xc++ -fsyntax-only -
# no error -- libstdc++ is installed ^

$ echo "#include <cstddef>" | clang -stdlib=libc++ -xc++ -fsyntax-only -
<stdin>:1:10: fatal error: 'cstddef' file not found
    1 | #include <cstddef>
      |          ^~~~~~~~~
1 error generated.
# error: libc++ is not installed ^
```

It seems we should be able to do the same using include-what-you-use?

Reply:
Note, only for C++ standard library, libc is hairier.

Reply:
Where I'm expecting to use this is in tests for standard library symbols. In the specific case of libc++ and libstdc++ a check using `-stdlib=*` might work. I don't think it extends to the libc's and not familiar enough with whether it would handle other STL's.

I suspect it also depends on clang having been compiled with appropriate support (i.e. knowing the paths to the different headers).

For IWYU though, if the right `-isystem` was passed in I think it would support being pointed at any libc/libc++ headers.

You're right though - I should be passing in all the IWYU test arguments... and that breaks my plan to try and cache the macros :(

Will have to have a think.

Reply:
I have some systematic ideas for caching, so let's forget about that for now

Reply:
I still have a feeling this is checking the wrong thing. Your C++ stdlib feature checks "does include-what-you-use use this standard library for this test run?".

I would have expected it to check is "is this standard library available on the system?" -- which will let you do stuff like:
```
// IWYU_REQUIRES: has-cxx-stdlb(libc++)
// IWYU_ARGS: -stdlib=libc++ -I .
```

I guess you could use the current feature for stuff like:
```
// IWYU_XFAIL: using-cxxstdlib(libc++)
// IWYU_ARGS: -I .
```

Reply:
"is this standard library available on the system" is a test that we can only apply for `libc++` and `libstdc++`, as those are the only things that clang is explicitly aware of.

I'm more interested in the test suite supporting the general case where someone has other libc or C++ standard library headers available. The IWYU test suite should be able to run against those headers.

For the standard library tests, I expect the same tests to apply to all libraries. And we'd just mark a few with `IWYU_XFAIL: using-cstdlib(macos)`. I don't think we want tests selecting a library using `IWYU_ARGS`. While it allows you to execute `libc++` and `libstdc++` tests in one run, it means duplicating tests for each standard library we want to support. Using the same files to test all libraries forces us to align IWYU behaviour on all implementations.

So someone could try and run:

    ./run_iwyu_tests.py   // run test suite with default libs
    IWYU_EXTRA_ARGS="-stdlib=libc++" ./run_iwyu_tests.py  // test clangs libc++
    IWYU_EXTRA_ARGS="-stdlib=libstdc++" ./run_iwyu_tests.py  // test clangs libstdc++
    IWYU_EXTRA_ARGS="--nostdlib -I /work/hpx -I /work/musl" ./run_iwyu_tests.py // test local HP C++ and musl
    IWYU_EXTRA_ARGS="--nostdlib -I /work/libcu++ -I /work/uclibc" ./run_iwyu_tests.py // test local Nvidia C++ and uclibc

Not that these libraries necessarily work together - that's up to the user to supply appropriate include paths and #defines in `IWYU_EXTRA_ARGS`.

Reply:
Thanks for the extended rationale, that makes a lot of sense to me.

I think there might be a case for `IWYU_REQUIRES: <some standard library>` + `-stdlib=that library` + `IWYU_XFAIL` for stuff like repro cases for bugs that only surface with particular standard libraries. It's always nicer to reduce to the language use that triggers the bug, but that usually takes quite a bit of effort.

But I see what you want to do here, and that also seems perfectly valid.

Reply:
Yes, we might want `IWYU_REQUIRES` + forcing the library for certain tests - those wouldn't behave in the above scenario, but I think that's fine.

FYI, I've started a [branch](/kilroyd/include-what-you-use/tree/cppref_tests) trying to generate some basic tests from cppreference. I think I want to allow manually adding `IWYU_XFAIL: using-libc()` in those and have the generator preserve any `IWYU_XFAIL` on updates. It already feels fiddly with macros. Functions feel doable. Templates... we'll have to see. Anyway more work required.